File: java/armstrong-numbers/src/main/java/ArmstrongNumbers.java
class ArmstrongNumbers {
	boolean isArmstrongNumber(int numberToCheck) {
		// if (numberToCheck == 0)
		// 	return true;
		//
		// int n = numberToCheck;
		// int len = (int)Math.ceil(Math.log10(numberToCheck));
		// int sum = 0;
		// while (n > 0) {
		// 	sum += Math.pow(n % 10, len);
		// 	n /= 10;
		// }
		// return sum == numberToCheck;

		String digits = String.valueOf(numberToCheck);
		int len = digits.length();
		int sum = digits
			.chars()
			.parallel()
			.map(c -> Character.getNumericValue(c))
			.map(d -> (int) Math.pow(d, len))
			.sum();
		return sum == numberToCheck;
	}
}

File: scala/armstrong-numbers/src/main/scala/ArmstrongNumbers.scala
import scala.math.floor
import scala.math.log10
import scala.math.pow
import scala.collection.mutable.ListBuffer

object ArmstrongNumbers {
  def isArmstrongNumber(n: Int): Boolean = n == n.armstrongSum
}

extension (n: Int)
  def width: Int = 1 + (if (n == 0) then 0 else floor(log10(n)).toInt)

  def digits: List[Int] =
    var m = n
    val ds = ListBuffer[Int]()
    while m > 0 do
      ds += m % 10
      m = m / 10
    ds.toList
    
  def armstrongSum: Int =
    val w = n.width
    n.digits.map {pow(_, w).toInt}.sum
